osquery/distributed/distributed.cpp

```diff
@@ -466,4 +513,34 @@ std::string hashQuery(const std::string& query) {
 uint64_t denylistDuration() {
   return FLAGS_distributed_denylist_duration;
 }
+
+SQL monitor_nonnumeric(const std::string& name, const std::string& query) {
+  // Snapshot the performance and times for the worker before running.
+  auto pid = std::to_string(PlatformProcess::getCurrentPid());
+  auto r0 = SQL::selectFrom({"resident_size", "user_time", "system_time"},
+                            "processes",
+                            "pid",
+                            EQUALS,
+                            pid);
+
+  using namespace std::chrono;
+  auto t0 = steady_clock::now();
+  Config::get().recordQueryStart(name);
+  SQL sql(query, true);
+
+  // Snapshot the performance after, and compare.
+  auto t1 = steady_clock::now();
+  auto r1 = SQL::selectFrom({"resident_size", "user_time", "system_time"},
+                            "processes",
+                            "pid",
+                            EQUALS,
+                            pid);
+  if (r0.size() > 0 && r1.size() > 0) {
+    // Always called while processes table is working.
+    uint64_t size = sql.rows().size();
+    Config::get().recordQueryPerformance(
+        name, duration_cast<milliseconds>(t1 - t0).count(), size, r0[0], r1[0]);
```

Comment:
A couple of things here:
1. This function also sets the scheduled executing query as not executing anymore, for denylisting purposes.
You can see it here:
https://github.com/osquery/osquery/blob/61e02f3026fbea77e94969cb142668fdf8448756/osquery/config/config.cpp#L1072-L1077

I would say that it has to be moved outside of that function and put only in the `launchQuery` logic of the scheduler.

2. Can't the name of a distributed query collide with the name of a scheduled query?

Reply:
That's fair. I can move the QueryPerformance objects into the instance of `Distributed` from `Config`. I'm not sure if this is going to be called from several different threads but I'll add a mutex to be safe.